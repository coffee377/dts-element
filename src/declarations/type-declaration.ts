import * as ts from 'typescript';
import {IType} from '../collections';
import {any_type} from '../constants';
import {ElementKind} from '../constants';
import {create_element, IElement, IElementOptions} from '../element';
import {transform} from '../transform';
import {create_type_parameters} from '../utils';
import {IGenericDeclaration} from './generic-declaration';

export interface ITypeDeclarationOptions extends IElementOptions {
  name: string;
  export?: boolean;
  generics?: IGenericDeclaration[];
  type?: IType;
}

export interface ITypeDeclaration
  extends IElement<ElementKind.TypeDeclaration>, ITypeDeclarationOptions {}

export const create_type_declaration = (options: ITypeDeclarationOptions): ITypeDeclaration => ({
  ...create_element(ElementKind.TypeDeclaration),
  ...options,
});

// tslint:disable:ter-indent

export const transform_type_declaration = (element: ITypeDeclaration, path: IElement<any>[]) => {
  const type_declaration = ts.createTypeAliasDeclaration(
    /* name            */ element.name,
    /* typeParameters  */ create_type_parameters(element.generics, path),
    /* type            */ transform(element.type || any_type, path) as ts.TypeNode,
  );
  if (element.export === true) {
    type_declaration.modifiers = [ts.createToken(ts.SyntaxKind.ExportKeyword)] as ts.NodeArray<ts.Modifier>;
  }
  return type_declaration;
};
