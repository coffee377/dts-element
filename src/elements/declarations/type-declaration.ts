import {any_type} from '../../constants';
import {emit_export} from '../../helpers/emit-export';
import {emit_generics} from '../../helpers/emit-generics';
import {Stack} from '../../stack';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {Type} from '../type';
import {GenericType} from '../types/generic-type';

export interface ITypeDeclarationRequiredParameters {
  name: string;
}

export interface ITypeDeclarationOptionalParameters {
  generics: GenericType[];
  target: Type;
  export: boolean;
}

export class TypeDeclaration
    extends Declaration<ITypeDeclarationRequiredParameters, ITypeDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & ITypeDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      generics: [],
      target: any_type,
      export: false,
    });
  }

  public _emit_raw(stack: Stack): string {
    const {name, generics} = this.parameters;
    const generic = emit_generics(generics, stack);
    const target = this.parameters.target.emit(stack);
    const an_export = emit_export(this.parameters.export, stack);
    return `${an_export}type ${name}${generic} = ${target};`;
  }

}
