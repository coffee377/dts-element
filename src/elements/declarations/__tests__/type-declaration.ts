jest.unmock('../type-declaration');

import {TypeDeclaration} from '../type-declaration';

describe('#emit()', () => {
  it('should return correctly', () => {
    expect(new TypeDeclaration({name: 'A'}).emit()).toMatchSnapshot();
  });
});
