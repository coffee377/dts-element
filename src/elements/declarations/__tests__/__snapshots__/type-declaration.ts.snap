// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#emit() should return correctly 1`] = `"[emit-jsdoc][emit-export]type A[emit-generics] = [BasicType any];"`;
